BindingAddress is customized for unix socket urls #10782 test data is added

mehmetcankamar opened the pull request:
fix for #10782

mehmetcankamar committed:
BindingAddress is customized for unix socket urls #10782 test data is added.

mehmetcankamar committed:
binding address, windows specific tests are moved another new method.

mehmetcankamar committed:
accidently deleted a test case is retrieved.

mjohn commented:
@dotnet-bot test macOS 10.13 please

mjohn commented:
@dotnet-bot help

mjohn commented:
@dotnet-bot test this please

davidfowl commented on src/Http/Http/src/BindingAddress.cs:
```diff
@@ -88,7 +89,15 @@ public static BindingAddress Parse(string address)
-                pathDelimiterStart = address.IndexOf(":", schemeDelimiterEnd + UnixPipeHostPrefix.Length, StringComparison.Ordinal);
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    pathDelimiterStart = address.IndexOf(":", schemeDelimiterEnd + UnixPipeHostPrefix.Length + 2, StringComparison.Ordinal);
```
Add a comment explaining the + 2.

Also it might be a bit cleaner like this:

```C#
var unixPipeHostPrefixLength = UnixPipeHostPrefix.Length;
if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
{
   // Windows has drive letters (c:)
   unixPipeHostPrefixLength  += 2;
}

pathDelimiterStart = address.IndexOf(":", schemeDelimiterEnd + unixPipeHostPrefixLength , StringComparison.Ordinal);
```

mehmetcankamar committed:
added coments and code is more cleaner now.

mehmetcankamar committed:
OSSkipCondition attribute moved under InlineData

mehmetcankamar committed:
skip reason is added to tests

mehmetcankamar committed:
BindingAddressTests for Windows specific were not windows specific, misspelled attribute changed from "Theory" to "ConditionalTheory"

davidfowl reviewed (approved):
The change looks great. Our parsing is wrt error handling in the bad cases though.

davidfowl reviewed (dismissed):
Ah! Actually we need a test that actually does the binding does function. To do that, you can add a test here:

https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs

It can be identical but instead of using the ListenDomainSocket API directly, you can do UseUrls with this syntax.

mehmetcankamar committed:
UnixPipePath had bug in windows. extra "/" character at the beginning is removed.

mjohn commented:
thanks for much detailed test scenario. I found a bug about UnixPipePath, now it's fixed. Currently, functional test couldn't succeed in my local, it gaves me error "-4092 EACCES permission denied", I'll try to resolve it or I may run tests on a console app then move it to test project.

mehmetcankamar committed:
functional tests for unixdomainsocket with url is completed

mjohn commented:
got fresh pull, restore and build, error is disappeared, I can run functional tests. It's much harder than I think. Unfortunately, .NET Core HttpClient doesn't support unix endpoint socket. I tried a hacky way to test without parsing raw http response. If anything I did wrong or could have been done better, please free feel to correct me.

mehmetcankamar committed:
unnecessary empty lines are removed.

davidfowl commented on src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs:
```diff
@@ -115,6 +116,93 @@ async Task EchoServer(ConnectionContext connection)
+
+                    using (var socket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified))
+                    {
+                        await socket.ConnectAsync(new UnixDomainSocketEndPoint(path));
```
Why is this code different from the test above? Can you make it match exactly?

mjohn replied:
I think, a URL which like "http://unix:/c:/temp/pipe.socket" should support HTTP, that's why I removed EchoServer. In addition to that, is there any way to implement nonhttp middleware without using listen endpoint? 

I'm sorry if I lack information about this domain.

davidfowl replied:
Ah, thats fine. It doesn't need to for this specific case.

halter73 commented on src/Http/Http/src/BindingAddress.cs:
```diff
@@ -88,7 +97,14 @@ public static BindingAddress Parse(string address)
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    // Windows has drive letters and volume separator (c:)
+                    unixPipeHostPrefixLength += 2;
```
Should we verify that the characters between `schemeDelimiterEnd`, and `schemeDelimiterEnd + unixPipeHostPrefixLength` actually matches `[A-Za-z]:` like we expect? What happens now someone tried to use a relative path to a single letter file in the PWD?

mjohn replied:
Good point, we should definitely verify it. I realized my assumptions were missing. Theoretically, we may also open an endpoint on a network shared folder like "\\COMPUTER-NAME\test". I'll research more.

mjohn replied:
Now, I added a length check and also it's checked that socket path is rooted or not on all operating systems.

mehmetcankamar committed:
TestUnixDomainSocketWithUrl test is now a little bit simpler

mehmetcankamar committed:
- Added a new validation for unix socket urls. Socket path must be rooted.
- Unix socket url parsing tests in UrlsAreParsedCorrectly are seperated based on OS.
- Socket path which has lower than 2 characters in Windows was causing unexpected exception. Fixed and new test is added.

anurse commented:
@mjohn looks like there are some conflicts here since we've moved up to 5.0, can you rebase on the latest code?

halter73 commented on src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs:
```diff
@@ -115,6 +117,67 @@ async Task EchoServer(ConnectionContext connection)
+                        var httpRequest = Encoding.ASCII.GetBytes("GET / HTTP/1.1\r\nHost:\r\n\r\n");
+                        await socket.SendAsync(httpRequest, SocketFlags.None);
+
+                        await socket.ReceiveAsync(readBuffer, SocketFlags.None);
```
@mjohn Sorry for the delay. This change looks really good, and we'd like to merge it soon. I have a few changes I'd like to make to this test to make it less flaky and less likely to hang.

The above line could return an incomplete read buffer leading to flaky tests. Can call ReceiveAsync in a loop until the expected amount of data is read or 0 is returned like the [above test](https://github.com/aspnet/AspNetCore/pull/12112/files#diff-a69a4830c53b497fe9f2eb25a73247ddL96-L103)?

The above test should check if ReceiveAsync returned 0 and exit the loop but doesn't. If you want to fix that at the same time great. If not, we can clean that up ourselves later. Every time we're awaiting a Task or ValueTask in a test, we should also be using the `.DefaultTimeout()` extension methods.

mjohn replied:
@halter73 Thanks for your comment, I tried to make your suggestion but ReceiveAsync didn't return 0. It hangs forever and receives timeout error. I'll push the branch and if I get over this error I can also fix the above test you mentioned.

I'm not sure whether the http respone size will vary. I also read http 1.1 specifications to determine response but chunked transfer encoding isn't easy as i thought.

halter73 replied:
In this case, we wouldn't expect ReceiveAsync to actually ever return 0. Since you're making an `HTTP/1.1` request without a `Connection: close` header, Kestrel assumes it's a keepalive request. The easiest fix is probably just add the  `Connection: close` to the request.

mehmetcankamar committed:
added DefaultTimeout extension method to async calls.
TestUnixDomainSocketWithUrl had incomplete http requests.

halter73 commented on src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs:
```diff
@@ -114,6 +116,73 @@ async Task EchoServer(ConnectionContext connection)
+                    {
+                        await socket.ConnectAsync(new UnixDomainSocketEndPoint(path)).DefaultTimeout();
+   
+                        var httpRequest = Encoding.ASCII.GetBytes("GET / HTTP/1.1\r\nHost:\r\n\r\n");
```
The following change should get you the 0-length read you're looking for.

```suggestion
                        var httpRequest = Encoding.ASCII.GetBytes("GET / HTTP/1.1\r\nHost:\r\nConnection: close\r\n\r\n");
```

mjohn replied:
Thank you I really appreciate your hel!. it's working like a charm.

halter73 commented on src/Servers/Kestrel/test/FunctionalTests/UnixDomainSocketsTests.cs:
```diff
@@ -114,6 +116,73 @@ async Task EchoServer(ConnectionContext connection)
+                            if (bytesReceived <= 0) break;
+                        }
+
+                        var httpResponse = Encoding.ASCII.GetString(readBuffer);
```
Nit:

```suggestion
                        var httpResponse = Encoding.ASCII.GetString(readBuffer, 0, read);
```

mehmetcankamar committed:
fixed infinite hang on TestUnixDomainSocketWithUrl tests

mehmetcankamar committed:
added a check to received bytes so it can break loops in TestUnixDomainSocket test

anurse commented:
@halter73 how's this one going? Ready to merge soon?

halter73 commented:
/azp run AspNetCore-ci

halter73 commented:
/azp run AspNetCore-ci

halter73 commented:
@anurse I think this is good to merge. All the AspNetCore-ci builds passed after a retry, but for some reason the GitHub PR status isn't being properly updated.

anurse commented:
> but for some reason the GitHub PR status isn't being properly updated.

I had the same problem earlier... I'll double check that's the case and force merge if so.
